src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityDataAccessHatch.java

```diff
@@ -103,35 +116,73 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
-
-        for (int y = 0; y < rowSize; y++) {
-            for (int x = 0; x < rowSize; x++) {
-                int index = y * rowSize + x;
-                builder.widget(new SlotWidget(isExportHatch ? exportItems : importItems, index,
-                        88 - rowSize * 9 + x * 18, 18 + y * 18, true, !isExportHatch)
-                                .setBackgroundTexture(GuiTextures.SLOT));
-            }
-        }
-        return builder.bindPlayerInventory(entityPlayer.inventory, GuiTextures.SLOT, 7, 18 + 18 * rowSize + 12)
-                .build(getHolder(), entityPlayer);
+    public boolean shouldOpenUI() {
+        return !isCreative;
     }
 
     @Override
-    protected boolean openGUIOnRightClick() {
-        return !this.isCreative;
+    public @NotNull ModularPanel buildUI(MetaTileEntityGuiData guiData, PanelSyncManager panelSyncManager,
+                                         UISettings settings) {
+        int rowSize = (int) Math.sqrt(getInventorySize());
+        panelSyncManager.registerSlotGroup("slots", rowSize);
+
+        Widget<?> recipeLogo = GTGuiTextures.getLogo(getUITheme())
+                .asWidget()
+                .align(Alignment.BottomRight)
+                .tooltipBuilder(tooltip -> {
+                    if (recipes.isEmpty()) {
+                        tooltip.addLine(IKey.lang("gregtech.machine.data_access_hatch.no_recipes"));
+                    } else {
+                        tooltip.addLine(IKey.lang("gregtech.machine.data_access_hatch.recipes"));
+                        tooltip.spaceLine(2);
+                    }
+
+                    Set<ItemStack> itemsAdded = new ObjectOpenCustomHashSet<>(ItemStackHashStrategy.comparingAll());
+                    for (Recipe recipe : recipes) {
+                        ItemStack output = recipe.getOutputs().get(0);
+                        if (itemsAdded.add(output)) {
+                            tooltip.add(new ItemDrawable(output));
+                            tooltip.space();
+                            tooltip.addLine(IKey.str(output.getDisplayName()));
+                        }
+                    }
+                });
+
+        return GTGuis.createPanel(this, 176, 18 + 18 * rowSize + 94)
+                .child(IKey.lang(getMetaFullName())
+                        .asWidget()
+                        .pos(5, 5))
+                .child(Flow.row()
+                        .top(18)
+                        .margin(7, 0)
+                        .coverChildrenHeight()
+                        .child(new Grid()
+                                .height(rowSize * 18)
+                                .minElementMargin(0, 0)
+                                .minColWidth(18).minRowHeight(18)
+                                .alignX(0.5f)
+                                .mapTo(rowSize, rowSize * rowSize, index -> new ItemSlot()
+                                        .slot(SyncHandlers.itemSlot(importItems, index)
```

Comment:
These maybe should have max stack size set to one. The data sticks would only stack if they have the exact same item researched, which would not be helpful for anything. So I think we can safely set the slot limit of these slots to a stack size of 1.